### ISSUE TITLE
 Rust program using OpenSSL 3.0.0 with FIPS Support enabled

### ISSUE THREAD

**Issue by vidhatha** [sfackler/rust-openssl#1924]
Hi, 
I am a newbie with openssl-rs and I am trying to get the FIPS mode enabled with OpenSSL 3.0.0 I have already looked at a couple of threads (https://github.com/sfackler/rust-openssl/issues/1404 , https://github.com/sfackler/rust-openssl/issues/1653, https://github.com/sfackler/rust-openssl/issues/1712), but this didn't help me with OpenSSL 3.0.0. These are the steps I have done:

1. Download OpenSSL 3.0.0
```
wget https://www.openssl.org/source/openssl-3.0.0.tar.gz
tar xvf openssl-3.0.0.tar.gz
cd openssl-3.0.0
```
2. Configure the build with the FIPS provider:
```
./config --prefix=/usr/local/ssl --openssldir=/usr/local/ssl fips
```
3. Build and install OpenSSL:
```
make
sudo make install
```
4. Updated the `/usr/local/ssl/openssl.cnf` file as mentioned in https://www.openssl.org/docs/man3.0/man7/fips_module.html
5. Set the OPENSSL_MODULES environment variable to point to the FIPS provider directory:
```
export OPENSSL_MODULES=/usr/local/ssl/lib/ossl-modules
```

Everything seemed to work fine. Now I am stuck on how to write a rust program using the fips provider. I tried using `openssl::fips::enable;` but I get an error

```
LD_LIBRARY_PATH=${OPENSSL_DIR}/lib ./rust_client 
Failed to enable FIPS mode: error:0F06D065:common libcrypto routines:FIPS_mode_set:fips mode not supported:crypto/o_fips.c:22:
```
which is correct, as this is deprecated for OpenSSL 3.0.0.

I think I am supposed to use [openssl::provider](https://docs.rs/openssl/latest/openssl/provider/index.html) but not exactly sure how to proceed. If anyone has successfully used OpenSSL 3.0.0 with FIPS enabled, any advice/ sample code would be greatly appreciated.

Thanks in advance!

**Comment by sfackler** [comment]
I believe something like

```rust
mem::forget(Provider::load(None, "fips")?);
```

At the very start of your program should work.

**Comment by sfackler** [comment]
The OpenSSL fips docs would be the best place to reference: https://www.openssl.org/docs/man3.0/man7/fips_module.html

**Comment by vidhatha** [comment]
yup. that was where I started off (mentioned in step 4 above :) ). But I wasn't able to proceed forward much with that. Maybe I am missing something there.

**Comment by vidhatha** [comment]
By the way the reason I wanted to use openSSL 3.0 is 1.1.1 is EOL on 11th September 2023 (https://www.openssl.org/blog/blog/2023/03/28/1.1.1-EOL/) so the recommended version is 3.0
